File: src/Game/Views/PlayerView.py
from pydantic import BaseModel
from typing import Dict, List, Optional
from Game.Components.GameState import GameState
from Game.Components.Card import Card
from Game.Components.Player import Player
from Game.Views.CardView import card_list_to_emoji_str_list
from Game.Systems.GameStateSystem import find_player, is_player_an_alpha
from Game.Modules.EventEnum import GameEventState


class PlayerView(BaseModel):
    name: str = ''
    uuid: str = ''
    can_start_game: bool = False
    my_turn: bool = False
    hosting: bool = False
    is_alpha: bool = False
    on_alpha_team: bool = False
    number_of_players: int = 0
    current_player: Optional[Player] = None
    leading_player: Optional[Player] = None
    winning_player_of_round: Optional[Player] = None
    player_list: List[Player] = list()
    player_hand: List[Card] = list()
    players_round_score: Dict[str, int] = dict()
    players_overall_score: Dict[str, int] = dict()
    game_event_state: GameEventState = GameEventState.NOT_AVAILABLE



def player_view_state(current_game_state: GameState, player_uuid: str) -> PlayerView:
    player_object = current_game_state.player_dict[player_uuid]
    player_view = PlayerView()
    player_view.uuid = str(player_object.uuid)
    player_view.name = str(player_object.name)
    player_view.game_event_state = current_game_state.game_event_state
    player_view.hosting = str(current_game_state.hosting_player.uuid) == str(player_object.uuid)
    player_view.is_alpha = is_player_an_alpha(current_game_state, player_object.uuid)
    player_view.number_of_players = len(current_game_state.player_order)
    player_view.players_round_score = current_game_state.players_round_score
    player_view.players_overall_score = current_game_state.players_overall_score
    player_view.can_start_game = current_game_state.can_start_game
    return player_view

File: frontend_code/src/test-utils/playerView.js
/* Fixtures mirroring the backend's PlayerView payload.
 *
 * Contract notes (see backend Game/Views/PlayerView.py):
 *   - card suit/rank arrive as enum NAMES ('HEART', 'ACE')
 *   - player levels arrive as enum VALUES (1..13, where ACE === 13)
 *   - game_event_state arrives as the enum's string value ('round-started')
 */

export const PLAYERS = [
    { uuid: 'uuid-alice', name: 'Alice' },
    { uuid: 'uuid-bob', name: 'Bob' },
    { uuid: 'uuid-carol', name: 'Carol' },
    { uuid: 'uuid-dave', name: 'Dave' },
    { uuid: 'uuid-erin', name: 'Erin' },
];

export const ME = PLAYERS[0];

export const card = (rank, suit) => ({ rank, suit });

/** A PlayerView with every field the frontend reads, defaulted to empty. */
export const playerView = (overrides = {}) => ({
    name: ME.name,
    uuid: ME.uuid,
    can_start_game: false,
    my_turn: false,
    hosting: false,
    is_alpha: false,
    on_alpha_team: false,
    number_of_players: PLAYERS.length,
    current_player: null,
    leading_player: null,
    winning_player_of_round: null,
    player_list: PLAYERS,
    disconnected_players: [],
    player_hand: [],
    players_round_score: {},
    players_overall_score: {},
    alpha_team_points: 0,
    defender_team_points: 0,
    my_team_points: 0,
    game_event_state: 'waiting-for-player-to-join',
    game_code: 'below-adopt-havoc',
    declare_trump: { rank: null, suit: null },
    cards_in_active_pile: [],
    leading_hand_of_subround: [],
    kitty_size: 0,
    my_level: 1,
    player_levels: PLAYERS.reduce((acc, p) => ({ ...acc, [p.uuid]: 1 }), {}),
    friend_calling_cards: [],
    num_friends_to_call: 0,
    revealed_friends: [],
    all_friends_found: false,
    round_winner_side: '',
    round_defender_points: 0,
    round_promotion_levels: 0,
    round_promoted_players: [],
    game_winner: '',
    events: [],
    ...overrides,
});

export const sessionInfo = (overrides = {}) => ({
    game_code: 'below-adopt-havoc',
    user_name: ME.name,
    user_uuid: ME.uuid,
    game_link: `/game/below-adopt-havoc/player/${ME.uuid}`,
    host: true,
    ...overrides,
});
